### DISCUSSION TITLE
How to allow an entity to share data with another entity

### DISCUSSION THREAD

**Discussion by gamemaker1** [Sunbird-RC/community#174] [Issues]
> As the title says.

## Example Scenario

For example, if I create two entities, a `Student` and a `Teacher`.

<details>

<summary> Click to see example schemas </summary>

**Student Schema**

> File name: `student.json`

```jsonc
{
  // This must be at the top of all schemas. See http://json-schema.org/understanding-json-schema/reference/schema.html#schema
  "$schema": "http://json-schema.org/draft-07/schema",
  // This is a schema object...
  "type": "object",
  // ...that declares the Student entity
  "properties": {
    "Student": { "$ref": "#/definitions/Student" }
  },
  "required": ["Student"],
  "definitions": {
    // The actual definition of the Student entity
    "Student": {
      // The path to the declaration in the schema entity
      "$id": "#/definitions/Student",
      // A Student is an object...
      "type": "object",
      // ...with the `name`, `email`, `phone` and `school` fields:
      "required": ["name", "email", "phone", "school"],
      "uniqueIndexFields": ["email"],
      "properties": {
        "name": { "type": "string" },
        "email": { "type": "string" },
        "phone": { "type": "string" },
        "school": { "type": "string" },
        "role": { "type": "string" }
      }
    }
  },
  "_osConfig": {
    // The following fields are used to create a corresponding user in Keycloak,
    // the authentication service used by Sunbird RC.
    "ownershipAttributes": [
      {
        // The path to the field to consider the email of the entity
        "email": "/email",
        // The path to the field to consider the phone number of the entity
        "mobile": "/phone",
        // The path to the field to consider as unique ID of the entity
        "userId": "/email"
      }
    ],
  }
}
```

**Teacher Schema**

> File name: `teacher.json`

```jsonc
{
  // This must be at the top of all schemas. See http://json-schema.org/understanding-json-schema/reference/schema.html#schema
  "$schema": "http://json-schema.org/draft-07/schema",
  // This is a schema object...
  "type": "object",
  // ...that declares the Teacher entity
  "properties": {
    "Teacher": { "$ref": "#/definitions/Teacher" }
  },
  "required": ["Teacher"],
  "definitions": {
    // The actual definition of the Teacher entity
    "Teacher": {
      // The path to the declaration in the schema entity
      "$id": "#/definitions/Teacher",
      // A Teacher is an object...
      "type": "object",
      // ...with the `name`, `email`, `phone` and `school` fields:
      "required": ["name", "email", "phone", "school"],
      "uniqueIndexFields": ["email"],
      "properties": {
        "name": { "type": "string" },
        "email": { "type": "string" },
        "phone": { "type": "string" },
        "school": { "type": "string" },
        "role": { "type": "string" }
      }
    }
  },
  "_osConfig": {
    // The following fields are used to create a corresponding user in Keycloak,
    // the authentication service used by Sunbird RC.
    "ownershipAttributes": [
      {
        // The path to the field to consider the email of the entity
        "email": "/email",
        // The path to the field to consider the phone number of the entity
        "mobile": "/phone",
        // The path to the field to consider as unique ID of the entity
        "userId": "/email"
      }
    ],
  }
}
```

</details>

Now if any application, say the Registry Frontend, wanted to access the `Student`
entity's data, they would go through the OAuth-based consent flow, as detailed
[here](https://docs.sunbirdrc.dev/using-the-apis#requesting-access-to-an-entitys-data).

However, I have not been able to find anything regarding an entity requesting
access to another entity's data (here, the `Teacher` requesting access to the 
`Student` entity's data):

- How would the granting of consent and subsequent retrieval of data occur between entities?
- Can only specific fields be shared with another entity?

Could someone please point to anything that would help answer the above questions?

Thanks

**Comment by dileepbapat** [comment]
@gamemaker1  It is not clear to me about the context of Teacher requesting Student info, is it related to claim about marks or something like that? There should be underlying activity / record that needs student info right? If it is claim flow might look something like below, For any other custom process you might need a backend for frontend (BFF) in between to orchestrate the flow.

![diagram](https://www.plantuml.com/plantuml/png/PP1DJiGm38NtESMxG2-GnMZGI4Xi4EA7lPLu9XQf0wsZNX-Yg9JIRVm-VtuvpqQJrJ6xLwQGUC9mmYMJZ4-vtaGzdlchycnuiHfP3GzwBUvF0iDmMeTR-h6ON2MGIL7dLlOdULSjUUJIo4Ki8IJAcVJ6heqE-alHkQ5MFbXttaqBZtksa7OtWZJYxJEICImnGaEtI_OU-5zswWquGlutFJ1zVL-zVk0nzIwyz2TtPerrpDy0)

**Reply by gamemaker1** [reply]
This claim flow seems to satisfy the use case I had described in the OP of selectively sharing data between entities. Could you please point me to an example/share the APIs to go through this flow?

Thanks.
